Robust hsu D flash method

msaitta-mpr opened the pull request:
### Description of the Change

Providing inputs to CoolProp of D and U (or any of HSU) is usually stable, but sometimes can cause annoying issues. This is noted in issues #1965 and #2157 (as well as some of my personal projects). This PR attempts to solve these issues once and for all as an extension to the work done in PR #2173. This is done in 3 main changes:

* Add additional checks and attempts at solving the ancillary equations. Adds more damping attempts and a way to force a guess.
* Add limits to the extrapolated ancillary guesses. If the ancillary becomes very flat near the edges (e.g., near the triple point or critical point), it might not have a solution, but the extrapolated guess could be very far off. We limit the extrapolated guess so that the Akshasa solver can at least attempt something. 
* If all else fails, we give up on finding the saturation curve, and apply a Brent method directly to find T by iterating H/S/U. We do not know the phase, but at least we have everything else. 

### Benefits

We should now always get a solution for D_HSU inputs. 

### Possible Drawbacks

It is possible that one of our fallbacks does not provide the user with a phase, instead setting it to iphase_unknown. 

The fallback solver can be slow because of needing a ton of Brent iterations, after already doing thousands of other iterations trying to find the saturation curve (which is now allowed to go to many of thousands of evaluations). However, a slow solution is almost always better than no solution. 

### Verification Process

I ran large sweeps of inputs to ensure that no errors were observed. I also integrated the results with another code and it seems to behave acceptably. I made custom test codes for issues #1965 and #2157. See the attached C++ code. 

[coolPropTester.zip](https://github.com/CoolProp/CoolProp/files/10314853/coolPropTester.zip)

### Applicable Issues

Closes #1965
Closes #2157

msaitta-mpr committed:
Increase robustness of saturation ancilliary invert

The SaturationAncillaryFunction::invert method is used to provide
guesses for several flash routines. However, if the inversion fails,
the entire flash routine will also fail. This commit makes the
routine more robust. Because it requires so many arguments, they
have been wrapped into their own struct.

msaitta-mpr committed:
Add limits to exapolated result in saturation invert

The exrapolating secant solver is needed to prevent issues when the
Brent solver cannot bound the provided density, but in some cases
the ancillary function has no solutio and the extrapolation goes
very far out of range. This commit adds a limit to the result.

msaitta-mpr committed:
More stability for out of bounds saturation guesses

Perform more clipping to prevent the ancilliary curve
solution from giving us a bad starting point.

msaitta-mpr committed:
Add ultimate fallback for HSU_D flash

Finding the saturation curve can still sometimes fail, despite our
best efforts. However, we can simply brute force our way to a
temperature solution through Brent's method. This is slow, and we do
not know the phase, but it is better than nothing.

ibell commented:
This is an impressive PR, addressing a particular point that has caused me a lot of pain.  Thank you very much.  Are there any points to discuss before merging?

@jowr Do you want to take a look?

msaitta-mpr committed:
Remove struct brace initializers

GCC does not seem to support brace initializers for structs with
default values. This commit removes the brace initializations, which
also makes the style more consistent throughout the code.

jowr commented:
Hi - I just surfaced again after a very busy period. I do not see any problems, but I haven't checked in detail. @msaitta-mpr  - I guess this PR is still relevant?

msaitta-mpr commented:
Both of the referenced issues are still open, so I assume that it is still a needed change. However, I have not attempted to rebase in the last 2 years, so I am not sure if there will be any unforseen merge issues.

ibell commented:
I have added superancillary functions and was getting ready to address this issue for pure fluids. @msaitta-mpr do you want to team up to do that (merged with this PR)?

msaitta-mpr commented:
OK. I will take a look and see what I can do.

ibell commented:
I think this PR has been superceded by #2995. Only helps when superancillaries are available, but that is basically all pure fluids now. The speedup is meaningful, and the reliability is very good
